### PR TITLE
Generate fixtures from e2e tests

### PR DESCRIPTION
**Pull request by jasonLaster**
This is a fun experiment for generating fixture data from E2E tests

https://www.notion.so/replayio/Automating-fixtures-c4c0da5fd7984057bf941330de0533ff

There are some really nice benefits of this approach

1. it's easy to create fixtures which can be used in unit tests
2. it's easy to re-generate fixtures when they're invalidated
3. it's easy to check for invalidation in a unit test in ci

### Checklist
- [ ] Decide on strategy for where we store fixture JSON data.
- [x] Make test runner check if fixture data changes between runs and warn locally or fail on CI.
- [x] Add `manifest.json` support for specifying which integration tests should generate fixture data.

### Follow up
- [ ] Decide on strategy for invalidating fixtures when protocol changes. (Perhaps encode the version into the fixture for comparison?)

### CODE REVIEW

**Comment by bvaughn** [comment]
79c613a adds our first unit test built with one of the new fixtures. I didn't touch the `runTest` script but my test depends on your path changes for the fixtures :)

**Comment by bvaughn** [comment]
I went ahead and updated the test fixture logic to add support for the arg `--updateWebsocketLogs` to instruct the runner to update fixtures. By default, it should do this when running on GitHub but not when running locally.

**Comment by bvaughn** [comment]
> Typecheck / GraphQL schema types (pull_request) 

This CI job is a burden 😬

**Comment by bvaughn** [comment]
Sorry for the poor form of force-pushing to your branch. I don't think you're around or actively working on this and I just didn't want to clutter the commit log with my silliness :)

**Comment by bvaughn** [comment]
I'm really excited about this direction. IMO the biggest downside is that these fixtures look pretty big. I wonder how folks would feel about putting them into a sibling repo or a submodule so they don't bloat this repo?

The `breakpoints-01` test for example generates a fixture that's > 80k lines long.

**Comment by bvaughn** [comment]
While generating fixture data for some of the other tests, I discovered that I can't even run some of these locally (like `console_warp-01.js`) because of the error:
```
Error: Cannot create a string longer than 0x1fffffe8 characters
    at Object.slice (node:buffer:593:37)
    at Buffer.toString (node:buffer:811:14)
    at Object.readFileSync (node:fs:497:41)
    at readRecordingFile (/Users/bvaughn/Documents/git/replay/devtools/test/node_modules/@replayio/replay/src/main.js:34:13)
    at readRecordings (/Users/bvaughn/Documents/git/replay/devtools/test/node_modules/@replayio/replay/src/main.js:65:17)
    at Object.listAllRecordings (/Users/bvaughn/Documents/git/replay/devtools/test/node_modules/@replayio/replay/src/main.js:222:22)
    at upload (/Users/bvaughn/Documents/git/replay/devtools/test/src/getExample.js:75:20)
    at recordExample (/Users/bvaughn/Documents/git/replay/devtools/test/src/getExample.js:108:23)
    at async getExample (/Users/bvaughn/Documents/git/replay/devtools/test/src/getExample.js:130:30)
    at async runMatchingTests (/Users/bvaughn/Documents/git/replay/devtools/test/run.js:52:34) {
  code: 'ERR_STRING_TOO_LONG'
```

At first I thought maybe we'd broken something, but it's that way on master as well 🤪

File: test/src/runTest.js [hunk 1]
```diff
@@ -34,6 +50,10 @@ async function recordBrowser(state, test, testPath, browserName) {
     await browser.close();
   }
 
+  if (updateWebsocketLogs) {
```
Comment (jasonLaster) [inline test/src/runTest.js:53]: it'd also be nice to have a field in `manifest.json` for specifying if we should save a fixture

Comment (jasonLaster) [reply test/src/runTest.js:53]: We should compare the new and old fixture and warn locally and fail in CI.

File: src/devtools/client/webconsole/components/__tests__/WebConsoleAppSearch.test.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    // Initialize state using exported websocket messages,
+    // sent through the mock environment straight to socket parsing
+    fixtureData.forEach((message: any) => {
+      window.mockEnvironment?.sendSocketMessage(JSON.stringify(message));
+    });
+
+    // Give everything time to settle
+    await sleep(100);
+  });
```
Comment (jasonLaster) [inline src/devtools/client/webconsole/components/__tests__/WebConsoleAppSearch.test.tsx:61]: it would be nice to extract this sooner rather than later to a util which takes the name of the fixture.

**Review by jasonLaster — COMMENTED** [review]
Hoping to be able to tweak this later tonight :)

File: test/src/runTest.js (hunk 1 above)
Comment (bvaughn) [reply test/src/runTest.js:53]: > We should compare the new and old fixture and warn locally and fail in CI.

Originally I thought we'd just run tests and only complain if they failed but I like this suggestion better.

File: src/devtools/client/webconsole/components/__tests__/WebConsoleAppSearch.test.tsx (hunk 2 above)
Comment (bvaughn) [reply src/devtools/client/webconsole/components/__tests__/WebConsoleAppSearch.test.tsx:61]: Agreed.

Comment (bvaughn) [reply src/devtools/client/webconsole/components/__tests__/WebConsoleAppSearch.test.tsx:21]: Done!

File: src/devtools/client/webconsole/components/Search/ConsoleSearch.tsx [hunk 3]
```diff
@@ -48,18 +48,32 @@ export default function ConsoleSearch({}: Props) {
...
+      if (results.length === 1) {
+        summaryMsg = "1 result";
+      } else {
+        summaryMsg = `${index + 1} of ${results.length} results`;
+      }
+    } else {
+      summaryMsg = "No results found";
+    }
+  }
+
```
Comment (bvaughn) [inline src/devtools/client/webconsole/components/Search/ConsoleSearch.tsx:63]: Going to extract this change into its own PR (#6544) since it's only tangentially test related.

File: src/test/testUtils.tsx [hunk 4]
```diff
@@ -60,7 +63,46 @@ export const filterLoggingInTests = (
...
+
+  // Give everything time to settle
+  await new Promise(resolve => setTimeout(resolve, 100));
+
+  return store;
+}
+
 export const filterCommonTestWarnings = () => {
+  // Skip LoadedRegions debug messages
+  filterLoggingInTests(message => message === "LoadedRegions", "debug");
```
Comment (bvaughn) [inline src/test/testUtils.tsx:105]: These debug logs are kind of annoying so I added this (for now). Maybe we should always filter out _all_ `console.debug` logs from test output.

**Comment by bvaughn** [comment]
I added a checklist to the PR description that gather the comments above into a single place listing what's left before this draft is done. Thoughts?

**Comment by jasonLaster** [comment]
###  Add manifest.json support for specifying which integration tests should generate fixture data.

Added a `saveFixture` field which will keep us from writing unnecessary fixtures to disk.

### Make test runner check if fixture data changes between runs and warn locally or fail on CI.

Added a naive string equality check and quickly realized that the messages were very different. (pause ids, session ids, timings, ...). We might want to encode the protocol version into the fixture, but at that point it seems like this can be something we can do when we have more data points.

**Comment by bvaughn** [comment]
> We might want to encode the protocol version into the fixture, but at that point it seems like this can be something we can do when we have more data points.

Fair!

**Comment by bvaughn** [comment]
Okay. Jason and I chatted out of band and decided to land this as-is with a single test fixture, then we'll follow up with a cleanup (that Jason will write an issue for) that removes the fixture and instead saves a Replay recording id in Git. Then we can extract the "fixture" messages _from the Replay recording itself_ (and save it in a Git-ignored .cache directory for offline support).

**Comment by jasonLaster** [comment]
Here's a quick [gist](https://gist.github.com/jasonLaster/8f1ff65479c642c87300459491dce8b8) of what a replay ID based fixture approach would look like. We can reference this when we follow up
